### PR TITLE
make toSeq not use compiles feature

### PR DESCRIPTION
**Pull request by krux02**
If you think `when compiles(...)` is bad, this should be an improvement.

This code is from my own fork of Nim. But I had to adapt it to "modern" nim. Here `lent` caused problems and I had to include special casing to get rid of it. The rest if as in my old Nim fork.

### CODE REVIEW

File: lib/pure/collections/sequtils.nim [hunk 1]
```diff
@@ -774,48 +774,86 @@ template anyIt*(s, pred: untyped): bool =
...
+    result[i] = a[i]
+    inc i
+
+proc toSeqBuiltin[E: enum](t: typedesc[E]): seq[E] =
+  for it in t:
+    result.add it
+
+macro toSeqBuiltin(arg: iterator): untyped =
+  let typ = arg.getTypeInst
+  typ.expectKind nnkProcTy # only for closure iterators
```
Comment (saem) [inline lib/pure/collections/sequtils.nim:806]: Should this be guarded in the JS case?

Just thinking about closure iterators being unsupported there.

File: lib/pure/collections/sequtils.nim [hunk 2]
```diff
@@ -774,48 +774,86 @@ template anyIt*(s, pred: untyped): bool =
...
+  result.add nnkVarSection.newTree(nnkIdentDefs.newTree(tmpSym, resultTyp, newNimNode(nnkEmpty)))
+  result.add quote do:
+    for it in `arg`:
+      `tmpSym`.add it
+  result.add tmpSym
+
+template toSeqBuiltin[T: not proc](arg: T): untyped =
+  # fallback for ``IntSet`` and ``HashSet``. Explicit overload would
+  # be better, but this would require an additional import here.
+  var tmp: seq[typeof(items(arg), typeOfIter)]
```
Comment (saem) [inline lib/pure/collections/sequtils.nim:822]: Seeing as we're in sequtils, wouldn't it be appropriate to map this? 😁

File: lib/pure/collections/sequtils.nim [hunk 3]
```diff
@@ -774,48 +774,86 @@ template anyIt*(s, pred: untyped): bool =
...
+  for it in arg:
+    tmp.add it
+  tmp
+
+macro toSeqImpl(arg: typed): untyped =
+  let iteratorCall = arg[1]
+  let iteratorSym = iteratorCall[0]
+  if iteratorCall.len == 2 and iteratorSym.eqIdent("items") and iteratorSym.owner.owner.eqIdent("stdlib"):
```
Comment (saem) [inline lib/pure/collections/sequtils.nim]: Minor: having a let for iteratorGrandParent or something, would trim the length if clause and each term would read at the same conceptual level.

File: lib/pure/collections/sequtils.nim [hunk 4]
```diff
@@ -774,48 +774,86 @@ template anyIt*(s, pred: untyped): bool =
...
-      for x in iter3():
-        result.add(x)
-    else:
-      for x in iter2():
-        result.add(x)
-    result
+    var typ = iteratorCall.getTypeInst()
+    # `lent` is broken feature. and this special handling to get rid of lent is why it is broken.
+    if typ.kind == nnkBracketExpr and typ.len == 2 and typ[0].eqIdent("lent"):
+      # TODO this is not guaranteed to be acutally be the system `lent` symbol.
```
Comment (saem) [inline lib/pure/collections/sequtils.nim]:
```suggestion
      # TODO this is not guaranteed to be the actual system `lent` symbol.
```

**Review by saem — CHANGES_REQUESTED** [review]
Mostly the typo in the todo, but it's easier to follow than before.

**Comment by saem** [comment]
For the commit message could you update it?

Commit message guide is here: https://nim-works.github.io/nimskull/contributing.html#pr-and-commit-message-guidelines-commit-message-guidelines

tl;dr:

```
where: what? 50 chars or less

*why*? format to 72 chars width

Also consider writing about "how?" and/or more elaborate explanation of
"what?"
```

These are the sorts of audience/scenario we're thinking about:
- changelog technical writer
- early contributors that are learning from commits
- dev trying to debug an issue
- sometime trying to figure out why some decision got made

That sort of thing.

File: lib/pure/collections/sequtils.nim (hunk 2 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:822]: you mean literally use `sequtils.map`? Well maybe, why not. I'll see if it causes problems.

File: lib/pure/collections/sequtils.nim (hunk 3 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim]: I think I would rather have a separate proc like `isStdlib` and then use that to shorten the line.

File: lib/pure/collections/sequtils.nim (hunk 2 above)
Comment (saem) [reply lib/pure/collections/sequtils.nim:822]: Sure, I'm also just being a bit cute, so feel free to not do it.

File: lib/pure/collections/sequtils.nim [hunk 5]
```diff
@@ -457,6 +457,13 @@ iterator filter*[T](s: openArray[T], pred: proc(x: T): bool {.closure.}): T {.ef
     if pred(s[i]):
       yield s[i]
 
+iterator filter*[T](s: seq[T], pred: proc(x: T): bool {.closure.}): T {.effectsOf: pred.} =
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim:460]: Need docs for this one.

File: lib/pure/collections/sequtils.nim [hunk 6]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
+macro toSeqBuiltin(arg: iterator): untyped =
+  let typ = arg.getTypeInst
+  typ.expectKind nnkProcTy # only for closure iterators
+  let formalParams = typ[0]
+  formalParams.expectKind nnkFormalParams
+  formalParams.expectMinLen 1
+  let resultTyp = nnkBracketExpr.newTree(bindSym"seq", formalParams[0])
+  let tmpSym = genSym(nskVar)
+  result = nnkStmtListExpr.newTree
+  result.add nnkVarSection.newTree(nnkIdentDefs.newTree(tmpSym, resultTyp, newNimNode(nnkEmpty)))
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim:813]: Some styling to add clarity to the code.

```suggestion
  result.add:
    nnkVarSection.newTree:
      nnkIdentDefs.newTree(tmpSym, resultTyp, newEmptyNode())
```

File: lib/pure/collections/sequtils.nim [hunk 7]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
+  for it in t:
+    result.add it
+
+macro toSeqBuiltin(arg: iterator): untyped =
+  let typ = arg.getTypeInst
+  typ.expectKind nnkProcTy # only for closure iterators
+  let formalParams = typ[0]
+  formalParams.expectKind nnkFormalParams
+  formalParams.expectMinLen 1
+  let resultTyp = nnkBracketExpr.newTree(bindSym"seq", formalParams[0])
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim:810]: Wouldn't this explode spectacularly if generics are used?

File: lib/pure/collections/sequtils.nim [hunk 8]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
+  let tmpSym = genSym(nskVar)
+  result = nnkStmtListExpr.newTree
+  result.add nnkVarSection.newTree(nnkIdentDefs.newTree(tmpSym, resultTyp, newNimNode(nnkEmpty)))
+  result.add quote do:
+    for it in `arg`:
+      `tmpSym`.add it
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim:816]: `:` can be very ambiguous sometimes, so it's better to split it away from command call.

```suggestion
  result.add:
    quote do:
      for it in `arg`:
        `tmpSym`.add it
```

File: lib/pure/collections/sequtils.nim [hunk 9]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
+    # `lent` is broken feature. and this special handling to get rid of lent is why it is broken.
+    if typ.kind == nnkBracketExpr and typ.len == 2 and typ[0].eqIdent("lent"):
+      # TODO this is not guaranteed to be the acutal system `lent` symbol.
+      typ = typ[1]
+
+    result = quote do:
+      var tmp: seq[`typ`]
+      for it in `iteratorCall`:
+        tmp.add it
+      tmp
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim:851]: It might be better to rope `sugar.collect` in instead, its impl is not perfect but it can do all this type inference properly.

**Review by alaviss — CHANGES_REQUESTED** [review]

File: lib/pure/collections/sequtils.nim [hunk 10]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
-        result.add(x)
-    else:
-      for x in iter2():
-        result.add(x)
-    result
+    var typ = iteratorCall.getTypeInst()
+    # `lent` is broken feature. and this special handling to get rid of lent is why it is broken.
+    if typ.kind == nnkBracketExpr and typ.len == 2 and typ[0].eqIdent("lent"):
+      # TODO this is not guaranteed to be the acutal system `lent` symbol.
+      typ = typ[1]
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim:845]: You can use this to get the return type of any iterable expression:

```nim
template iterRetTyp(u: untyped): untyped =
  proc inferer(): auto {.genSym.} =
    for it in u:
      return it

  typeof(inferer())

iterator test[T](): T =
  yield default(T)

echo iterRetTyp(test[int]())
echo iterRetTyp(['a', 'b', 'c'].items)
```

With this you basically hand the job to the compiler, so it's always correct.

File: lib/pure/collections/sequtils.nim (hunk 6 above)
Comment (Vindaar) [reply lib/pure/collections/sequtils.nim:813]: Damn, I keep forgetting this is valid syntax 👀

File: lib/pure/collections/sequtils.nim (hunk 5 above)
Comment (beef331) [reply lib/pure/collections/sequtils.nim:460]: Should this not be `openArray[T]`?

File: lib/pure/collections/sequtils.nim (hunk 1 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:806]: not sure, can you provide an example that could cause problems?

File: lib/pure/collections/sequtils.nim (hunk 5 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:460]: an overload for `filter` with `openArray` already exists. It also already has the documentation needed. Not sure if Nimworks is different here, but IIRC sets of overloads that do the same thing don't all need a copy paste documentation string. Or do they now?

But yes, documentation needs to be added for this one. Not for what it does (obvious), but for why it is even necessary and why the `openArray` alone doesn't cut it. This is the situation:

```nim
for it in myiterator(expr()):
  discard
```

When myiterator takes an ``openArray`` argument, ``expr`` is reevaluated on each loop cycle. There are test cases that ensure on ``toSeq`` to not have side effects. In essence I am working around a Nim compiler bug here.

If you have suggestions on what I should write, I am open for suggestions. I don't know how do document these things.

File: lib/pure/collections/sequtils.nim (hunk 8 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:816]: Actually I've written a lot of code in this concise way (command call). I think it is very readable. Especially this exact pattern `result.add quote do:` is a pattern I've used a lot in the past. I think my opengl sandbox project is full of it.

Long story short, I'd prefer not not split the line here.

File: lib/pure/collections/sequtils.nim (hunk 7 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:810]: can you provide an example? I am asking, because also for me it was quite some time age since I originally wrote this rewrite of `toSeq`, it is just recent that I contributed it to nimskull.

File: lib/pure/collections/sequtils.nim [hunk 11]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
+    result = quote do:
+      var tmp: seq[`typ`]
+      for it in `iteratorCall`:
+        tmp.add it
+      tmp
+
+proc genCallToSeqImpl(arg: NimNode): NimNode =
+  newCall(bindSym"toSeqImpl",
+    nnkForStmt.newTree(
+      ident"x", arg, newStmtList()))
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim:856]: Some structuring to improve clarity:

```suggestion
  newCall(bindSym"toSeqImpl"):
    nnkForStmt.newTree(ident"x", arg):
      newStmtList()
```

File: lib/pure/collections/sequtils.nim (hunk 8 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:816]: It's more about how the compiler sometimes bind the `do:` to the wrong callee, which I tripped on a few times while developing CPS.

It's fine if you left it as-is in this case.

File: lib/pure/collections/sequtils.nim (hunk 7 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:810]: I couldn't get that part to trip up, which is great, but I caused some other code to fail, though.

File: lib/pure/collections/sequtils.nim (hunk 1 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:806]: I'd say it's not a concern.

File: lib/pure/collections/sequtils.nim (hunk 10 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:845]: I think you can use `typeKind` to identify the type of `lent` and/or `var` to skip as well.

File: lib/pure/collections/sequtils.nim (hunk 5 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:460]: You should document that it is a workaround for a compiler bug (please also file an issue for that bug) and hyperlink it to the other iterator, ie.

> Overload of \<link this\>`filter(openArray[T]) iterator` to workaround a compiler bug...

Comment (alaviss) [reply lib/pure/collections/sequtils.nim:460]: With that said, it seems like you are not using this new iterator in this PR. Can you split this part out to a separate PR instead?

File: lib/pure/collections/sequtils.nim [hunk 12]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
-      for it in s2:
-        result[i] = it
-        i += 1
-      result
-  else:
-    var result: seq[OutType] = @[]
-    for it in s:
-      result.add(it)
-    result
+template toSeqBuiltin[IX, T](arg: array[IX, T]): seq[T] = @arg
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim]: Wouldn't the `openArray` version perform the job of this one?

File: lib/pure/collections/sequtils.nim [hunk 13]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
-    var result = newSeq[typeof(iter2)](iter2.len)
-    for x in iter2:
-      result[i] = x
-      inc i
-    result
+template toSeqBuiltin[T](arg: openArray[T]): seq[T] = @arg
+
+template toSeqBuiltin(arg: string): seq[char] = @arg
+
+template toSeqBuiltin[T](arg: seq[T]): seq[T] = arg
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim]: Should be covered by the `openArray` version, no?

File: lib/pure/collections/sequtils.nim [hunk 14]
```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
...
+      tmp
+
+proc genCallToSeqImpl(arg: NimNode): NimNode =
+  newCall(bindSym"toSeqImpl",
+    nnkForStmt.newTree(
+      ident"x", arg, newStmtList()))
+
+macro toSeqImplSingleSym(arg: typed): untyped =
+  if arg.kind != nnkSym:
+    error "identifier not unique", arg # arg is symChoice
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim:860]: I think it's better to generate a call for it and have the compiler ambiguous call handling trigger instead, which would give a better error message.

**Review by alaviss — CHANGES_REQUESTED** [review]
A test case that fails with this new implementation:

```nim
import std/sequtils

iterator test[T](): T =
  yield default(T)

echo toSeq(test[int])
```

Error:

```
test.nim(6, 12) Error: type mismatch: got <iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}>
but expected one of:
iterator items(a: cstring): char
  first type mismatch at position: 1
  required type for a: cstring
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items(a: string): char
  first type mismatch at position: 1
  required type for a: string
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items[IX, T](a: array[IX, T]): T
  first type mismatch at position: 1
  required type for a: array[IX, T]
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items[T: Ordinal](s: Slice[T]): T
  first type mismatch at position: 1
  required type for s: Slice[items.T]
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items[T: char](a: openArray[T]): T
  first type mismatch at position: 1
  required type for a: openArray[T: char]
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items[T: enum and Ordinal](E: typedesc[T]): T
  first type mismatch at position: 1
  required type for E: typedesc[T: enum and Ordinal]
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items[T: not char](a: openArray[T]): lent2 T
  first type mismatch at position: 1
  required type for a: openArray[T: not char]
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items[T](a: seq[T]): lent2 T
  first type mismatch at position: 1
  required type for a: seq[T]
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items[T](a: set[T]): T
  first type mismatch at position: 1
  required type for a: set[T]
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
iterator items[T](xs: iterator (): T): T
  first type mismatch at position: 1
  required type for xs: iterator (): T{.closure.}
  but expression 'test' is of type: iterator (): int{.inline, noSideEffect, gcsafe, locks: 0.}
  Calling convention mismatch: got '{.inline.}', but expected '{.closure.}'.

expression: items(test)
```

File: lib/pure/collections/sequtils.nim (hunk 9 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:851]: well that would inject another (big?) import  into `sequtils`. Not sure if that is a good idea.

**Comment by krux02** [comment]
@alaviss 
> ```nim
> test[int]
> ```

well you get the problem because `test[int]` is actually not a call to an itearator in the first place. It isn't a call at all. You should write `toSeq(test[int]())` at least.

**Comment by krux02** [comment]
@alaviss 
> With that said, it seems like you are not using this new iterator in this PR. Can you split this part out to a separate PR instead?

That iterator is used in the tests. I added it to make the tests not fail. As I said. I am working around a compiler bug.

File: lib/pure/collections/sequtils.nim (hunk 5 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:460]: I've added a comment. I also added elsewhere why this part can't be put into another PR. I hope this answers your questions.

File: lib/pure/collections/sequtils.nim (hunk 10 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:845]: I've written extensive macros. I've written Nim to glsl compilers like in OpenGL sandbox. I've never ever used `typeKind` for anything, and would like to keep it that way.

File: lib/pure/collections/sequtils.nim (hunk 11 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:856]: I know where you come from. Is there a style guide that enforces this for the compiler? I am asking, because I am also a lisp programmer, I am really really used to the style that I've been put in here. In other words, I'd rather not do it, if that is allowed.

File: lib/pure/collections/sequtils.nim (hunk 14 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:860]: Sorry I don't know what you mean, or if it is possible. Again it is really long ago since I wrote it, but knowing myself, I certainly would have done it that way, if I would have seen it feasible.

File: lib/pure/collections/sequtils.nim (hunk 10 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:845]: You are going into `typed` land and the best way to be correct is to ask the compiler. It's one of the parts in typed macro to be solid-ish and trustable.

File: lib/pure/collections/sequtils.nim (hunk 11 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:856]: We haven't formally adopted a style yet, but this is not Lisp and I'd prefer that we use something more idiomatic for Nim, that is, embrace spacing and use block structures where appropriate.

File: lib/pure/collections/sequtils.nim (hunk 14 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:860]: I mean to just `genCallToSeqImpl(newCall(arg))` it, which passes the burden downstream for the compiler to handle.

File: lib/pure/collections/sequtils.nim [hunk 15]
```diff
@@ -457,6 +457,15 @@ iterator filter*[T](s: openArray[T], pred: proc(x: T): bool {.closure.}): T {.ef
     if pred(s[i]):
       yield s[i]
 
+iterator filter*[T](s: seq[T], pred: proc(x: T): bool {.closure.}): T {.effectsOf: pred.} =
+  # this overload of `fileter` is written to work around this bug:
```
Comment (alaviss) [inline lib/pure/collections/sequtils.nim]:
```suggestion
  # this overload of `filter` is written to work around this bug:
```

File: lib/pure/collections/sequtils.nim (hunk 10 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:845]: It is not about typed vs untyped. It is about the fact that `typeKind` feels 100% redundant to `getTypeInst.kind`. It is not obvious why a separate proc with its own enum type is necessary here, when all of this information is already present in the ast.

Maybe you are right in terms of getting rid of the `lent` flag (And I will change my PR accordingly). But that doesn't change my opinion on the redundancy of this feature. After all this information is contained in the type ast, therefore it could be eleminated.

File: lib/pure/collections/sequtils.nim (hunk 11 above)
Comment (saem) [reply lib/pure/collections/sequtils.nim:856]: No style guide yet, but I have to admit after CPS I grew to have a significant appreciation for the approach @alaviss is suggesting. Admittedly, it was really off putting at first, I don't have a strong opinion. I just fiddle with indentation and alignment until it feels like the visual cues are drawing attention to the right things.

File: lib/pure/collections/sequtils.nim (hunk 13 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim]: if you look closely, the `seq` and the `array` implementation do slightly differ. On `seq` I don't generate an `@` because the type is already correct (I know `@` on `seq` isn't supposed to to actually have an effect as well).

File: lib/pure/collections/sequtils.nim (hunk 10 above)
Comment (krux02) [reply lib/pure/collections/sequtils.nim:845]: I just tested `typeKind`, it returns `ntyUnused2`. In other words, no `typeKind` can't be used here.

Comment (krux02) [reply lib/pure/collections/sequtils.nim:845]: @alaviss regarding `iterRetTyp` I've tested it. and it works. But I'd rather not use it here. The major reason for it is, I am in macro land, I should be able to just ask the compiler "gimme the return type of this iterator", I should not need to rely of crutches like you suggested to get my type. And I do it, I ask the compiler to get the type, but the compiler also puts in garbage into my answer and that is a problem that needs to be fixed. In order for problems to be fixed, they need to be visible. That is what I am doing here, I make a problem visible. I just can't fix it within this PR.

File: lib/pure/collections/sequtils.nim (hunk 13 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim]: The non-seq versions should all be covered by the `openArray` overload so you should be able to trim most of the fat there.

File: lib/pure/collections/sequtils.nim (hunk 10 above)
Comment (alaviss) [reply lib/pure/collections/sequtils.nim:845]: Can you check if comparing directly with `bindSym"lent"` works then? That'd be a lot more robust.

**Review by alaviss — APPROVED** [review]
Looks mostly fine to me!

Thanks for the contribution @krux02 🎉

bors r+

**Comment by saem** [comment]
Yay, congrats on the first PR, @krux02.

You helped unblock me on my early compiler commits in Nim and I haven't forgotten that. Super happy that you are contributing here. 🎉❤️

**Comment by krux02** [comment]
> Yay, congrats on the first PR, @krux02.
> 
> You helped unblock me on my early compiler commits in Nim and I haven't forgotten that. Super happy that you are contributing here. tadaheart

Well but I forgot that I helped you. Honestly I can't remember. But thank you for reminding me.
